File: scripts/update-data.js
#!/usr/bin/env node

/**
 * Data Ingestion Script for UAE Attacks Monitor
 * 
 * This script updates the Supabase database with attack data.
 * It requires the SUPABASE_SERVICE_ROLE_KEY which has write permissions.
 * 
 * Usage:
 *   node scripts/update-data.js                    # Update today's data
 *   node scripts/update-data.js --date 2024-03-15  # Update specific date
 *   node scripts/update-data.js --bulk             # Bulk update from JSON file
 */

import { createClient } from '@supabase/supabase-js';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import readline from 'readline';

// Get __dirname equivalent for ES modules
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Environment variables
const SUPABASE_URL = process.env.SUPABASE_URL || process.env.VITE_SUPABASE_URL;
const SUPABASE_SERVICE_ROLE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!SUPABASE_URL || !SUPABASE_SERVICE_ROLE_KEY) {
    console.error('❌ Missing required environment variables:');
    if (!SUPABASE_URL) console.error('   - SUPABASE_URL or VITE_SUPABASE_URL');
    if (!SUPABASE_SERVICE_ROLE_KEY) console.error('   - SUPABASE_SERVICE_ROLE_KEY');
    process.exit(1);
}

// Initialize Supabase client with service role key
const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY, {
    auth: {
        autoRefreshToken: false,
        persistSession: false
    }
});

// Parse command line arguments
const args = process.argv.slice(2);
const isHelp = args.includes('--help') || args.includes('-h');
const isBulk = args.includes('--bulk');
const dateIndex = args.indexOf('--date');
const specificDate = dateIndex !== -1 && args[dateIndex + 1] ? args[dateIndex + 1] : null;

if (isHelp) {
    console.log(`
UAE Attacks Monitor - Data Update Script

Usage:
  node scripts/update-data.js [options]

Options:
  --help, -h           Show this help message
  --date YYYY-MM-DD    Update data for a specific date
  --bulk               Bulk update from public/data.json
  
Examples:
  node scripts/update-data.js                    # Update today's data
  node scripts/update-data.js --date 2024-03-15  # Update March 15, 2024
  node scripts/update-data.js --bulk             # Import all data from JSON

Environment Variables Required:
  SUPABASE_URL or VITE_SUPABASE_URL
  SUPABASE_SERVICE_ROLE_KEY
`);
    process.exit(0);
}

/**
 * Format date to match the JSON format (e.g., "Mar 15")
 */
function formatDateShort(date) {
    const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 
                   'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
    const d = new Date(date);
    const month = months[d.getMonth()];
    const day = String(d.getDate()).padStart(2, '0');
    return `${month} ${day}`;
}

/**
 * Convert short date format to full date (e.g., "Mar 15" -> "2024-03-15")
 */
function parseShortDate(shortDate, year = new Date().getFullYear()) {
    const months = {
        'Jan': '01', 'Feb': '02', 'Mar': '03', 'Apr': '04',
        'May': '05', 'Jun': '06', 'Jul': '07', 'Aug': '08',
        'Sep': '09', 'Oct': '10', 'Nov': '11', 'Dec': '12'
    };
    
    const [monthName, day] = shortDate.split(' ');
    const month = months[monthName];
    return `${year}-${month}-${day.padStart(2, '0')}`;
}

/**
 * Update or insert a single day's data
 */
async function updateSingleDay(dateStr, uav, cruise, ballistic) {
    try {
        // Convert to short date format if it's a full date
        const shortDate = dateStr.length > 6 ? formatDateShort(dateStr) : dateStr;
        
        // First, check if record exists
        const { data: existing, error: selectError } = await supabase
            .from('attacks')
            .select('*')
            .eq('date', shortDate)
            .single();

        let result;
        if (selectError && selectError.code === 'PGRST116') {
            // Record doesn't exist, insert it
            result = await supabase
                .from('attacks')
                .insert({
                    date: shortDate,
                    uav: uav,
                    cruise: cruise,
                    ballistic: ballistic
                });
        } else if (existing) {
            // Record exists, update it
            result = await supabase
                .from('attacks')
                .update({
                    uav: uav,
                    cruise: cruise,
                    ballistic: ballistic
                })
                .eq('date', shortDate);
        } else {
            throw selectError;
        }

        const { error } = result;
        if (error) {
            console.error(`❌ Error updating ${shortDate}:`, error.message);
            return false;
        }

        console.log(`✅ Updated ${shortDate}: UAV=${uav}, Cruise=${cruise}, Ballistic=${ballistic}`);
        return true;
    } catch (err) {
        console.error(`❌ Unexpected error updating ${dateStr}:`, err);
        return false;
    }
}

/**
 * Bulk update from JSON file
 */
async function bulkUpdate() {
    try {
        const jsonPath = path.join(__dirname, '..', 'public', 'data.json');
        const jsonData = JSON.parse(fs.readFileSync(jsonPath, 'utf8'));
        
        console.log(`📊 Starting bulk update of ${jsonData.length} records...`);
        
        let successCount = 0;
        let errorCount = 0;
        
        for (const record of jsonData) {
            // The JSON already has short date format, use it directly
            const success = await updateSingleDay(
                record.date,
                record.uav,
                record.cruise,
                record.ballistic
            );
            
            if (success) {
                successCount++;
            } else {
                errorCount++;
            }
            
            // Add a small delay to avoid rate limiting
            await new Promise(resolve => setTimeout(resolve, 100));
        }
        
        console.log(`\n📈 Bulk update complete:`);
        console.log(`   ✅ Success: ${successCount}`);
        console.log(`   ❌ Errors: ${errorCount}`);
        
    } catch (err) {
        console.error('❌ Failed to read or parse data.json:', err);
        process.exit(1);
    }
}

/**
 * Interactive update for today or specific date
 */
async function interactiveUpdate(date) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    
    const ask = (question) => new Promise(resolve => {
        rl.question(question, resolve);
    });
    
    try {
        console.log(`\n📅 Updating data for: ${date}`);
        console.log('Enter the number of attacks (press Enter for 0):\n');
        
        const uav = parseInt(await ask('UAV attacks: ') || '0');
        const cruise = parseInt(await ask('Cruise missile attacks: ') || '0');
        const ballistic = parseInt(await ask('Ballistic missile attacks: ') || '0');
        
        console.log(`\n📊 Summary for ${date}:`);
        console.log(`   UAV: ${uav}`);
        console.log(`   Cruise: ${cruise}`);
        console.log(`   Ballistic: ${ballistic}`);
        
        const confirm = await ask('\nConfirm update? (y/n): ');
        
        if (confirm.toLowerCase() === 'y') {
            await updateSingleDay(date, uav, cruise, ballistic);
        } else {
            console.log('❌ Update cancelled');
        }
        
    } finally {
        rl.close();
    }
}

/**
 * Main execution
 */
async function main() {
    if (isBulk) {
        await bulkUpdate();
    } else {
        const targetDate = specificDate || new Date().toISOString().split('T')[0];
        await interactiveUpdate(targetDate);
    }
    
    process.exit(0);
}

// Run the script
main().catch(err => {
    console.error('❌ Fatal error:', err);
    process.exit(1);
});

File: scripts/update-data.py
#!/usr/bin/env python3
"""
UAE Attacks Monitor - Data Update Script (Python Version)

This script provides a simple way to update attack data in Supabase.
It can be run locally or in GitHub Actions.

Requirements:
    pip install supabase python-dotenv

Usage:
    python scripts/update-data.py                      # Interactive mode
    python scripts/update-data.py --date 2024-03-15 --uav 10 --cruise 2 --ballistic 5
    python scripts/update-data.py --csv data.csv      # Update from CSV file
"""

import os
import sys
import argparse
import csv
from datetime import datetime, date
from typing import Optional, Dict, List

try:
    from supabase import create_client, Client
    from dotenv import load_dotenv
except ImportError:
    print("❌ Missing required packages. Please install:")
    print("   pip install supabase python-dotenv")
    sys.exit(1)

# Load environment variables
load_dotenv()

class DataUpdater:
    def __init__(self):
        """Initialize Supabase client with service role key"""
        url = os.getenv('SUPABASE_URL') or os.getenv('VITE_SUPABASE_URL')
        service_key = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
        
        if not url or not service_key:
            print("❌ Missing required environment variables:")
            if not url:
                print("   - SUPABASE_URL or VITE_SUPABASE_URL")
            if not service_key:
                print("   - SUPABASE_SERVICE_ROLE_KEY")
            sys.exit(1)
        
        self.supabase: Client = create_client(url, service_key)
        print("✅ Connected to Supabase")
    
    def update_single_day(self, date_str: str, uav: int, cruise: int, ballistic: int) -> bool:
        """Update or insert data for a single day"""
        try:
            response = self.supabase.table('attacks').upsert({
                'date': date_str,
                'uav': uav,
                'cruise': cruise,
                'ballistic': ballistic
            }, on_conflict='date').execute()
            
            print(f"✅ Updated {date_str}: UAV={uav}, Cruise={cruise}, Ballistic={ballistic}")
            return True
            
        except Exception as e:
            print(f"❌ Error updating {date_str}: {e}")
            return False
    
    def update_from_csv(self, csv_file: str) -> None:
        """Bulk update from CSV file"""
        if not os.path.exists(csv_file):
            print(f"❌ CSV file not found: {csv_file}")
            return
        
        success_count = 0
        error_count = 0
        
        with open(csv_file, 'r') as f:
            reader = csv.DictReader(f)
            for row in reader:
                try:
                    date_str = row['date']
                    uav = int(row.get('uav', 0))
                    cruise = int(row.get('cruise', 0))
                    ballistic = int(row.get('ballistic', 0))
                    
                    if self.update_single_day(date_str, uav, cruise, ballistic):
                        success_count += 1
                    else:
                        error_count += 1
                        
                except (KeyError, ValueError) as e:
                    print(f"❌ Skipping invalid row: {row} - {e}")
                    error_count += 1
        
        print(f"\n📊 CSV import complete:")
        print(f"   ✅ Success: {success_count}")
        print(f"   ❌ Errors: {error_count}")
    
    def interactive_update(self, date_str: Optional[str] = None) -> None:
        """Interactive mode for updating data"""
        if not date_str:
            date_str = date.today().isoformat()
        
        print(f"\n📅 Updating data for: {date_str}")
        print("Enter the number of attacks (press Enter for 0):\n")
        
        try:
            uav = int(input("UAV attacks: ") or "0")
            cruise = int(input("Cruise missile attacks: ") or "0")
            ballistic = int(input("Ballistic missile attacks: ") or "0")
            
            print(f"\n📊 Summary for {date_str}:")
            print(f"   UAV: {uav}")
            print(f"   Cruise: {cruise}")
            print(f"   Ballistic: {ballistic}")
            
            confirm = input("\nConfirm update? (y/n): ")
            
            if confirm.lower() == 'y':
                self.update_single_day(date_str, uav, cruise, ballistic)
            else:
                print("❌ Update cancelled")
                
        except ValueError:
            print("❌ Invalid input. Please enter numbers only.")
        except KeyboardInterrupt:
            print("\n❌ Update cancelled")
    
    def get_recent_data(self, days: int = 7) -> None:
        """Display recent data from the database"""
        try:
            response = self.supabase.table('attacks')\
                .select("*")\
                .order('date', desc=True)\
                .limit(days)\
                .execute()
            
            print(f"\n📊 Last {days} days of data:")
            print("Date       | UAV | Cruise | Ballistic")
            print("-----------|-----|--------|----------")
            
            for record in response.data:
                print(f"{record['date']} | {record['uav']:3d} | {record['cruise']:6d} | {record['ballistic']:9d}")
                
        except Exception as e:
            print(f"❌ Error fetching data: {e}")


def main():
    parser = argparse.ArgumentParser(description='Update UAE Attacks Monitor data')
    parser.add_argument('--date', help='Date to update (YYYY-MM-DD)')
    parser.add_argument('--uav', type=int, help='Number of UAV attacks')
    parser.add_argument('--cruise', type=int, help='Number of Cruise missile attacks')
    parser.add_argument('--ballistic', type=int, help='Number of Ballistic missile attacks')
    parser.add_argument('--csv', help='Import data from CSV file')
    parser.add_argument('--show', type=int, metavar='DAYS', 
                      help='Show recent data (last N days)')
    
    args = parser.parse_args()
    updater = DataUpdater()
    
    if args.show:
        updater.get_recent_data(args.show)
    elif args.csv:
        updater.update_from_csv(args.csv)
    elif args.date and args.uav is not None and args.cruise is not None and args.ballistic is not None:
        updater.update_single_day(args.date, args.uav, args.cruise, args.ballistic)
    else:
        updater.interactive_update(args.date)


if __name__ == '__main__':
    main()